### ISSUE TITLE
Auto-discovery of pachd ADDRESS (at least for minikube)

### ISSUE THREAD

**Issue by gabrielgrant** [pachyderm/pachyderm#3089]
**What is the goal / desired outcome?**

To have it be as easy as possible to connect to pachd. Right now running `pachctl` commands on Minikube requires either manually setting `ADDRESS` or running `pachctl port-forward`. Neither of these should be necessary.

**What is your proposal for a feature to solve this?**

My understanding is that right now we default to connecting to localhost:30650 if no `ADDRESS` env var is set. If, instead, we defaulted to using the host of the currently-active kubernetes cluster, I believe that should make local minikube and unsecured cloud deployments work out of the box.

Using kubectl in bash:
```
current_context=$(kubectl config view --output jsonpath='{.current-context}')
kube_uri=$(kubectl config view --output jsonpath="{.clusters[?(@.name == \"${current_context}\")].cluster.server}")
kube_authority=${kube_uri#*//}  # get from first "//" onwards
kube_host=${kube_authority%:*}  # get from last ":" backwards
ADDRESS=${kube_host}:30650
```

For secured cloud instances, using either ADDRESS or `port-forward` would still be necessary.

(@suneeta-mall had [a suggestion a few weeks ago](https://github.com/pachyderm/pachyderm/issues/1881#issuecomment-402354085) that might also make this work better for a wider variety of clusters, but I'm not totally sure I understand the proposal -- care to explain a bit more?)

**If there is a way to accomplish this today via workaround, what does that require?**

manually setting `ADDRESS` or running `pachctl port-forward`

**Comment by suneeta-mall** [comment]
@gabrielgrant Sorry I missed your question earlier ..  

Lets start with background:  My view at the time of comment was biased with assumption that Kubernetes cluster will be shared with Pachyderm alongside of any other services/kube resources. I appreciate this wont always be true. 

Lets assume that its true for a bit, then finding A-record for API server is really as simple as digging or 'discovering' IP of the API server hostname (via network tools/libraries). Since Pachyderm port is accessible from all nodes then going via any of the master node IP will be fair. So if we just dig the cluster api server for IPs and chose one - that will simply work with how you are proposing to setup ADDRESS already. 

This is how I get past it for example:
```
dig +short api.$(kubectl config current-context) | while read mip ; do 
        export ADDRESS=$mip:30650
    echo Possible host $ADDRESS
done
```

Now, this clearly safely assumes that your current kube context is where pachyderm is deployed. But this can be an option e.g. `--context`  or  `--use-context` etc. 

In essence, I was proposing that instead of asking user to explicitly set this (which I find annoying), it would be fairly better user experience (something easily achievable) to  'discover' this value and set it internally.  

So then its just about using argument  `--context`` (to pachctl) to override default (current-context) for cases where users are using non current-context Kube cluster as Pachyderm environment. The save really is by default pachctl just works without having to explicitly set anything (or provide context for special case). 

**TLDR** Use context to 'discover' the ip of API server and by default set it to ADDRESS as you are already proposing. The context can default to current.  This leaves us with port-forward being an optional thing thats useful for dashboard or should user prefer. I hope that clarifies a bit. Happy to be corrected if I oversaw anything

**Comment by gabrielgrant** [comment]
@suneeta-mall thanks for the thorough explanation -- it does sound like this should work for a fair number of cloud-based clusters :+1:

**Comment by msteffen** [comment]
@ysimonson Do you see a way around https://github.com/pachyderm/pachyderm/issues/3089#issuecomment-416727290?

I actually didn't even realize this issue was still open, though if you see a solution then maybe that was for the best

**Comment by ysimonson** [comment]
Not yet. I self-assigned to dig deeper at a later point. Happy to unassign if you think it's best, though.

**Comment by ysimonson** [comment]
A couple of points in favor of (1):

* It's of course generally good to make the initial experience as seamless as possible, and I would guess most users kicking the tires would be using minikube.
* It would default to not using the proxy, which is presumably slower.

That said, my first point is presumptuous. Plus I could see this change being really confusing for existing users who are accustomed to the current behavior.

As another alternative, would it be feasible to instead first try `localhost:30650`, and failing that, discover the host of the actively running cluster?

**Comment by jdoliner** [comment]
A better approach to this might be having `pachctl` automatically port-forward for each command. This is, according to user comment, how `helm` talks to `tiller` and it is, again according to user comment, quite reliable.

**Comment by ysimonson** [comment]
I like that. The only downside I can think of is that the user will default to going through the proxy, which will be slower.

**Comment by ysimonson** [comment]
I'm not aware of any reason we can't/shouldn't change those defaults, but that's a very different ticket from this. Maybe it makes sense to close this out and open a new one?

**Comment by gabrielgrant** [comment]
Sounds good, opened as #3417
